### PR TITLE
SNOW-1794369 Catalog API

### PR DESCRIPTION
**Pull request by sfc-gh-aalam**
1. Which Jira issue is this PR addressing? Make sure that there is an accompanying issue to your PR.

   

   Fixes SNOW-1794369

2. Fill out the following pre-review checklist:

   - [ ] I am adding a new automated test(s) to verify correctness of my new code
      - [ ] If this test skips Local Testing mode, I'm requesting review from @snowflakedb/local-testing
   - [ ] I am adding new logging messages
   - [ ] I am adding a new telemetry message
   - [ ] I am adding new credentials
   - [ ] I am adding a new dependency
   - [ ] If this is a new feature/behavior, I'm adding the Local Testing parity changes.
   - [x] I acknowledge that I have ensured my changes to be thread-safe. Follow the link for more information: [Thread-safe Developer Guidelines](https://docs.google.com/document/d/162d_i4zZ2AfcGRXojj0jByt8EUq-DrSHPPnTa4QvwbA/edit#bookmark=id.e82u4nekq80k)

3. Please describe how your code solves the related issue.

   This PR adds the catalog API.

### CODE REVIEW

File: tests/integ/test_catalog.py [hunk 1]
```diff
@@ -0,0 +1,464 @@
...
+    assert (
+        len(
+            catalog.list_tables(
+                pattern="does_not_exist_*", database=temp_db1, schema=temp_schema1
+            )
+        )
+        == 0
+    )
+
+    print(f"list tables {temp_db1} {temp_schema1}")
```
Comment (sfc-gh-aalam) [inline tests/integ/test_catalog.py:208]:
```suggestion
```

File: docs/source/snowpark/catalog.rst [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    Catalog.userDefinedFunctionExists
+    Catalog.user_defined_function_exists
+    Catalog.viewExists
+    Catalog.view_exists
+    .. Catalog.functionExists
+    .. Catalog.function_exists
+    .. Catalog.getFunction
+    .. Catalog.get_function
+    .. Catalog.listFunctions
+    .. Catalog.list_functions
```
Comment (sfc-gh-aalam) [inline docs/source/snowpark/catalog.rst:68]:
```suggestion
```

File: src/snowflake/snowpark/catalog.py [hunk 3]
```diff
@@ -0,0 +1,636 @@
...
+
+import snowflake.snowpark
+from snowflake.snowpark._internal.type_utils import convert_sp_to_sf_type
+from snowflake.snowpark.types import DataType
+
+
+class Column(NamedTuple):
+    name: str
+    datatype: str
+    nullable: bool
```
[the remark this answers is not in the document]
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/catalog.py:26]: I agree but I don't know how what is a suitable alternative here. Do you have any suggestions?

File: src/snowflake/snowpark/session.py [hunk 4]
```diff
@@ -686,6 +688,18 @@ def get_active_session(cls) -> Optional["Session"]:
...
+    @property
+    def catalog(self) -> Catalog:
+        """Returns the catalog object."""
+        if self._catalog is None:
+            if isinstance(self._conn, MockServerConnection):
+                self._conn.log_not_supported_error(
+                    external_feature_name="Session.catalog",
+                    raise_error=NotImplementedError,
+                )
+            self._catalog = Catalog(self)
```
[the remark this answers is not in the document]
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/session.py:700]: yeah. it similar to `Dataframe`

File: src/snowflake/snowpark/catalog.py [hunk 5]
```diff
@@ -0,0 +1,636 @@
...
+from snowflake.core.table import Table
+from snowflake.core.user_defined_function import UserDefinedFunction
+from snowflake.core.view import View
+
+import snowflake.snowpark
+from snowflake.snowpark._internal.type_utils import convert_sp_to_sf_type
+from snowflake.snowpark.types import DataType
+
+
+class Column(NamedTuple):
```
Comment (sfc-gh-yixie) [inline src/snowflake/snowpark/catalog.py:23]: `snowflake.core` also has Column class. Should we reuse that one?

Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/catalog.py:23]: let me check, maybe I can use that for my implementation.

Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/catalog.py]: thanks. updated the PR

File: src/snowflake/snowpark/catalog.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/catalog.py]: I used Yijun's suggestions and use python api's `TableColumn` for this.

**Review by sfc-gh-jrose — APPROVED** [review]

File: setup.py [hunk 6]
```diff
@@ -29,6 +29,7 @@
     # `protoc` < 3.20 is not able to generate protobuf code compatible with protobuf >= 3.20.
     "protobuf>=3.20, <6",  # Snowpark IR
     "tzlocal",  # Snowpark IR
+    "snowflake>=1.0.0, <2",  # Catalog
```
Comment (sfc-gh-aalam) [inline setup.py:32]: use snowflake.core

File: src/snowflake/snowpark/catalog.py [hunk 7]
```diff
@@ -0,0 +1,623 @@
...
+    ) -> List[Database]:
+        """List databases in the current session.
+
+        Args:
+            pattern: the pattern of name to match. Defaults to None.
+        """
+        iter = self._root.databases.iter()
+        if pattern:
+            iter = filter(lambda x: re.match(pattern, x.name), iter)
+
```
Comment (sfc-gh-aalam) [inline src/snowflake/snowpark/catalog.py:110]: using UDF will give a better performance while providing api parity

File: src/snowflake/snowpark/catalog.py [hunk 8]
```diff
@@ -0,0 +1,659 @@
...
+        Args:
+            database: database name or ``Database`` object. Defaults to None.
+            pattern: the pattern of name to match. Defaults to None.
+        """
+        db_name = self._parse_database(database)
+        iter = self._root.databases[db_name].schemas.iter()
+        if pattern:
+            iter = filter(lambda x: re.match(pattern, x.name), iter)
+        return list(iter)
```
Comment (sfc-gh-aalam) [inline src/snowflake/snowpark/catalog.py:172]: DatabaseModel and SchemaModel don't have from_json methods so we cannot use udf filter for them.

File: src/snowflake/snowpark/catalog.py [hunk 9]
```diff
@@ -0,0 +1,659 @@
...
+            self._python_regex_udf = self._session.udf.register(python_regex_filter)
+
+    def _list_objects(
+        self,
+        *,
+        object_name: str,
+        object_class,
+        database: Optional[Union[str, Database]],
+        schema: Optional[Union[str, Schema]],
+        pattern: Optional[str],
```
Comment (sfc-gh-yixie) [inline src/snowflake/snowpark/catalog.py:115]: After double-thinking, I suggest we add the `like` parameter as well which is a good snowflake feature.

File: src/snowflake/snowpark/catalog.py [hunk 10]
```diff
@@ -0,0 +1,659 @@
...
+            database=database,
+            schema=schema,
+            pattern=pattern,
+        )
+
+    # get methods
+    def get_current_database(self) -> Database:
+        """Get the current database."""
+        current_db_name = self._session.get_current_database()
+        return self._root.databases[current_db_name]
```
[the remark this answers is not in the document]
Comment (sfc-gh-aalam) [reply src/snowflake/snowpark/catalog.py:289]: I added catalog.py to pytest checker. Such mistakes would be caught by it.

File: src/snowflake/snowpark/catalog.py [hunk 11]
```diff
@@ -0,0 +1,682 @@
...
+                return bool(re.match(pattern, input))
+
+            self._python_regex_udf = self._session.udf.register(python_regex_filter)
+
+    def _list_objects(
+        self,
+        *,
+        object_name: str,
+        object_class,
+        database: Optional[Union[str, Database]],
```
Comment (sfc-gh-yixie) [inline src/snowflake/snowpark/catalog.py:112]: For optional arguments, better give it a None as default value.

File: src/snowflake/snowpark/catalog.py [hunk 12]
```diff
@@ -0,0 +1,682 @@
...
+            db_name = model_obj.database_name
+            assert db_name is not None  # pyright
+            return db_name
+
+        if isinstance(database, str):
+            return database
+        if isinstance(database, Database):
+            return database.name
+        if database is None:
+            return self._session.get_current_database()
```
Comment (sfc-gh-yixie) [inline src/snowflake/snowpark/catalog.py:52]: get_current_database() could returns None as well.

File: src/snowflake/snowpark/catalog.py [hunk 13]
```diff
@@ -0,0 +1,682 @@
...
+            schema_name = model_obj.schema_name
+            assert schema_name is not None  # pyright
+            return schema_name
+
+        if isinstance(schema, str):
+            return schema
+        if isinstance(schema, Schema):
+            return schema.name
+        if schema is None:
+            return self._session.get_current_schema()
```
Comment (sfc-gh-yixie) [inline src/snowflake/snowpark/catalog.py:74]: Same here. get_current_schema could returns None.

File: src/snowflake/snowpark/catalog.py [hunk 14]
```diff
@@ -0,0 +1,682 @@
...
+            object_name="USER FUNCTIONS",
+            object_class=UserDefinedFunction,
+            database=database,
+            schema=schema,
+            pattern=pattern,
+            like=like,
+        )
+
+    # get methods
+    def get_current_database(self) -> Database:
```
Comment (sfc-gh-yixie) [inline src/snowflake/snowpark/catalog.py:307]: better return the database name

File: src/snowflake/snowpark/catalog.py [hunk 15]
```diff
@@ -0,0 +1,682 @@
...
+        """Get the current database."""
+        current_db_name = self._session.get_current_database()
+        return self._root.databases[current_db_name].fetch()
+
+    def get_current_schema(self) -> Schema:
```
Comment (sfc-gh-yixie) [inline src/snowflake/snowpark/catalog.py:312]: same as database, it returns string

**Review by sfc-gh-yixie — APPROVED** [review]

**Pull request merged**